subsys/net/l2/ethernet/ethernet.c

```diff
@@ -675,9 +675,24 @@ static int ethernet_send(struct net_if *iface, struct net_pkt *pkt)
 			case 0x60:
 				ptype = htons(NET_ETH_PTYPE_IPV6);
 				break;
-			case 0x40:
+			case 0x40: {
```

Comment:
I start to wonder, if it shouldn't be `net_context` responsibility to replace `net_pkt_family()` type with `AF_INET`/`AF_INET6` instead of passing `AF_PACKET` down for `IPPROTO_RAW` sockets/contexts. For the Ethernet layer it really makes no difference what is the origin of the IP packet, and currently we're duplicating even more code.

Reply:
Excellent comment, agree, didn't like the duplication as well but couldn't found better place. 

So, you mean to set the pkt family here instead? https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/net/ip/net_context.c#L1766

With the quick trial it seems to work. Need to do wider testing still, also with L2/ppp. where we could also remove the similar code accordingly.

Reply:
Yes, that'd be the place in the `net_context` where the packet family could be updated.

Reply:
OK, setting pkt family for IPv4 and IPv6 in net_context level and removed corresponding stuff from lower levels (ethernet and ppp). Now no additional code duplication is needed for doing ARP. Updated, PR description and subject accordingly.
Please, review.

Reply:
Unfortunately this change for this PR break the use cases of raw sckt because zephyr ip stack got involved more. Thus, fallbacking to how this PR was originally: https://github.com/zephyrproject-rtos/zephyr/pull/54664